### PR TITLE
introduce TAKE_FD() and TAKE_PTR() macros

### PR DESCRIPTION
**Pull request by poettering**
As suggested by @sourcejedi here:

https://github.com/systemd/systemd/pull/8532#discussion_r176246657

I think it's a good idea, makes our code shorter and a bit more readable (at least after you grokked the concept, but that should be quick enough)

### CODE REVIEW

File: src/resolve/resolved-dns-stub.c [hunk 1]
```diff
@@ -447,10 +447,7 @@ static int manager_dns_stub_udp_fd(Manager *m) {
                 return r;
 
         (void) sd_event_source_set_description(m->dns_stub_udp_event_source, "dns-stub-udp");
-        m->dns_stub_udp_fd = fd;
-        fd = -1;
-
-        return m->dns_stub_udp_fd;
+        return TAKE_FD(fd);
```
Comment (sourcejedi) [inline src/resolve/resolved-dns-stub.c]: The `return` cocci is problematic (and the same for TAKE_PTR).  Maybe don't include that part of the cocci file in the commit?

I think this hunk is supposed to be

```
m->dns_stub_udp_fd = TAKE_FD(fd);

return m->dns_stub_udp_fd;
```

File: src/resolve/resolved-dns-stub.c [hunk 2]
```diff
@@ -542,10 +539,7 @@ static int manager_dns_stub_tcp_fd(Manager *m) {
                 return r;
 
         (void) sd_event_source_set_description(m->dns_stub_tcp_event_source, "dns-stub-tcp");
-        m->dns_stub_tcp_fd = fd;
-        fd = -1;
-
-        return m->dns_stub_tcp_fd;
+        return TAKE_FD(fd);
```
Comment (sourcejedi) [inline src/resolve/resolved-dns-stub.c]: and here.  I think it was just a problem in this one file.

**Comment by poettering** [comment]
Yikes. Good catch! Fixed the two cases, and force pushed a new version. PTAL!

i also tweaked the coccinelle scripts a bit, so that we only match on local variables in that case, which successfully ensures that cases like that or not matched.

**Comment by sourcejedi** [comment]
Looking good :).

There's something else that always catches my eye - what's the purpose of close_nointr() _not_ reporting EINTR?  We don't worry about the caller incorrectly retrying for the other errors (EBADF, EIO).

EINTR means we were supposed to be doing some IO - so this was the last reference to the file description - and I think we _have_ to assume the kernel may not do any further IO to flush it in the background, otherwise that IO would stop an NFS from being unmounted, and that's only supposed to happen iff an fd is open on the NFS.  Shorter: -EINTR suggests that the IO required to complete the request has been aborted.

(I doubt this nasty corner gets fixed in Linux, if it's even possible to do so with current ABIs, but I don't think that would _add_ any reason to handle EINTR differently).

**Comment by poettering** [comment]
> EINTR means we were supposed to be doing some IO - so this was the last reference to the file description - and I think we have to assume the kernel may not do any further IO to flush it in the background, otherwise that IO would stop an NFS from being unmounted, and that's only supposed to happen iff an fd is open on the NFS. Shorter: -EINTR suggests that the IO required to complete the request has been aborted.

To my knowledge EINTR from close() doesn't mean the IO got aborted, but that we stopped waiting for it. Hence there's nothing wrong really: the fd is invalidated, and the IO will proceed just fine. The only thing is that something that is supposed to be synchronous now is asynchronous

**Comment by sourcejedi** [comment]
Right, maybe the function comment could hint towards that better.  Looking at current kernel source indeed disproves my contradiction.

Although the link I followed from the comment talks about NFS, since 2.6.25 NFS syscalls cannot be interrupted except to kill the process.  All syscalls on filesystem fds should now be uninterruptible.  This resolves historic issues with interrupting nfs close().  Interruptions can no longer stop you being notified that you ran out of disk quota, hooray :).

close() v.s. network sockets only blocks if you messed around with SO_LINGER, which is a can of worms generally best avoided.  I don't think it blocks for pipes.  Character devices like tty's can at least in principle have special-purpose ioctls like tc_drain(), to wait safely for your data before close().

**Pull request merged**